### ISSUE TITLE
redis script cache gets bloated after update to bull 2.0

### ISSUE THREAD

**Issue by prdn** [OptimalBits/bull#426]
Hello,

I'm experiencing a growing Lua script cache after updating to bull-2.0
After 20 days of uptime my redis Lua cache grown up to 15GB

Anyone experiencing same issue?

Thanks

**Comment by manast** [comment]
can you check what entries are there in the cache? there must be a spurious entry that is for some reason not being reused but created every time it is used.

**Comment by prdn** [comment]
@manast I've added some console logs to the script caching function in `lib/scripts.js`. The hashes seem all the same.
Just to be sure that we are on the same, the value that is growing fast is `used_memory_lua`
Reverting to v1.1.3 is working fine.

Do you have a preferred way to analyse the redis lua script cache?

**Comment by prdn** [comment]
@manast can it be that `_retryAtOnce` is using directly `eval` and not caching the script and using `evalsha` so it is accumulating the same script again and again every time there is a retry?

**Comment by manast** [comment]
eval is not caching the script, so don't think that is the reason of the problem. I am not aware of any other command in redis besides LOAD that caches scripts. So if you have time you could put a console.log just before calling load, and you should be able to see which script is being cached continuously: https://redis.io/commands/script-load

**Comment by prdn** [comment]
You're right. I added a console.log in my dev environment.
It looks everything normal but just a nodejs process (concurrency = 5) is bloating my redis instance.
On production I need to revert to 1.1.3. The main change seems moving to ioredis and that looks good. I will try to spend a bit of time on it in next days, since it's crucial for me to maintain `bull` as a dependency.

Thanks for helping

**Comment by manast** [comment]
The major new thing in 2.0 is the redlock algo, I would bet that is the root of the problem. @doublerebel  any ideas?

**Comment by prdn** [comment]
I think you're right @manast 
Is there an easy way to test the previous locking system?
Otherwise will set up a branch for it

**Comment by manast** [comment]
when the old locking system was replaced by redlock many other changes where performed, so it is not easy to replace it, and I do not think it is a good idea either. I will investigate when I have some time why the script cache is increasing, there could be an issue in the redlock dependency as well: https://github.com/mike-marcacci/node-redlock

**Comment by manast** [comment]
I have analised the code, and I cannot see anything in redlock or in bull that would currently cache a lot of scripts. In redlock it is actually running eval directly (which is not so performant, I would like to improve this in the future), and in bull, the code related to lock is not using any scripts it is relayed to redlock. If you put a console log here https://github.com/OptimalBits/bull/blob/master/lib/scripts.js#L46 outputing the hashes you should only see a few ones otherwise you should see a lot of them. 
Now, is this cache bloating creating problems? it could be some redis optimization that caches lua stuff but that is reused by actual keys if necessary, and therefore nothing to worry about?

**Comment by manast** [comment]
From redis documentation https://redis.io/commands/eval
```
Available since 2.6.0.
Time complexity: Depends on the script that is executed.
Introduction to EVAL
EVAL and EVALSHA are used to evaluate scripts using the Lua interpreter built into Redis starting from version 2.6.0.
The first argument of EVAL is a Lua 5.1 script. The script does not need to define a Lua function (and should not). It is just a Lua program that will run in the context of the Redis server.
The second argument of EVAL is the number of arguments that follows the script (starting from the third argument) that represent Redis key names. The arguments can be accessed by Lua using the KEYS global variable in the form of a one-based array (so KEYS[1], KEYS[2], ...).
All the additional arguments should not represent key names and can be accessed by Lua using the ARGV global variable, very similarly to what happens with keys (so ARGV[1], ARGV[2], ...).
The following example should clarify what stated above:
> eval "return {KEYS[1],KEYS[2],ARGV[1],ARGV[2]}" 2 key1 key2 first second
1) "key1"
2) "key2"
3) "first"
4) "second"
Note: as you can see Lua arrays are returned as Redis multi bulk replies, that is a Redis return type that your client library will likely convert into an Array type in your programming language.
It is possible to call Redis commands from a Lua script using two different Lua functions:
redis.call()
redis.pcall()
redis.call() is similar to redis.pcall(), the only difference is that if a Redis command call will result in an error, redis.call() will raise a Lua error that in turn will force EVAL to return an error to the command caller, while redis.pcall will trap the error and return a Lua table representing the error.
The arguments of the redis.call() and redis.pcall() functions are all the arguments of a well formed Redis command:
> eval "return redis.call('set','foo','bar')" 0
OK
The above script sets the key foo to the string bar. However it violates the EVAL command semantics as all the keys that the script uses should be passed using the KEYS array:
> eval "return redis.call('set',KEYS[1],'bar')" 1 foo
OK
All Redis commands must be analyzed before execution to determine which keys the command will operate on. In order for this to be true for EVAL, keys must be passed explicitly. This is useful in many ways, but especially to make sure Redis Cluster can forward your request to the appropriate cluster node.
Note this rule is not enforced in order to provide the user with opportunities to abuse the Redis single instance configuration, at the cost of writing scripts not compatible with Redis Cluster.
Lua scripts can return a value that is converted from the Lua type to the Redis protocol using a set of conversion rules.
Conversion between Lua and Redis data types
Redis return values are converted into Lua data types when Lua calls a Redis command using call() or pcall(). Similarly Lua data types are converted into the Redis protocol when a Lua script returns a value, so that scripts can control what EVAL will return to the client.
This conversion between data types is designed in a way that if a Redis type is converted into a Lua type, and then the result is converted back into a Redis type, the result is the same as the initial value.
In other words there is a one-to-one conversion between Lua and Redis types. The following table shows you all the conversions rules:
Redis to Lua conversion table.
Redis integer reply -> Lua number
Redis bulk reply -> Lua string
Redis multi bulk reply -> Lua table (may have other Redis data types nested)
Redis status reply -> Lua table with a single ok field containing the status
Redis error reply -> Lua table with a single err field containing the error
Redis Nil bulk reply and Nil multi bulk reply -> Lua false boolean type
Lua to Redis conversion table.
Lua number -> Redis integer reply (the number is converted into an integer)
Lua string -> Redis bulk reply
Lua table (array) -> Redis multi bulk reply (truncated to the first nil inside the Lua array if any)
Lua table with a single ok field -> Redis status reply
Lua table with a single err field -> Redis error reply
Lua boolean false -> Redis Nil bulk reply.
There is an additional Lua-to-Redis conversion rule that has no corresponding Redis to Lua conversion rule:
Lua boolean true -> Redis integer reply with value of 1.
Also there are two important rules to note:
Lua has a single numerical type, Lua numbers. There is no distinction between integers and floats. So we always convert Lua numbers into integer replies, removing the decimal part of the number if any. If you want to return a float from Lua you should return it as a string, exactly like Redis itself does (see for instance the ZSCORE command).
There is no simple way to have nils inside Lua arrays, this is a result of Lua table semantics, so when Redis converts a Lua array into Redis protocol the conversion is stopped if a nil is encountered.
Here are a few conversion examples:
> eval "return 10" 0
(integer) 10

> eval "return {1,2,{3,'Hello World!'}}" 0
1) (integer) 1
2) (integer) 2
3) 1) (integer) 3
   2) "Hello World!"

> eval "return redis.call('get','foo')" 0
"bar"
The last example shows how it is possible to receive the exact return value of redis.call() or redis.pcall() from Lua that would be returned if the command was called directly.
In the following example we can see how floats and arrays with nils are handled:
> eval "return {1,2,3.3333,'foo',nil,'bar'}" 0
1) (integer) 1
2) (integer) 2
3) (integer) 3
4) "foo"
As you can see 3.333 is converted into 3, and the bar string is never returned as there is a nil before.
Helper functions to return Redis types
There are two helper functions to return Redis types from Lua.
redis.error_reply(error_string) returns an error reply. This function simply returns a single field table with the err field set to the specified string for you.
redis.status_reply(status_string) returns a status reply. This function simply returns a single field table with the ok field set to the specified string for you.
There is no difference between using the helper functions or directly returning the table with the specified format, so the following two forms are equivalent:
return {err="My Error"}
return redis.error_reply("My Error")
Atomicity of scripts
Redis uses the same Lua interpreter to run all the commands. Also Redis guarantees that a script is executed in an atomic way: no other script or Redis command will be executed while a script is being executed. This semantic is similar to the one of MULTI / EXEC. From the point of view of all the other clients the effects of a script are either still not visible or already completed.
However this also means that executing slow scripts is not a good idea. It is not hard to create fast scripts, as the script overhead is very low, but if you are going to use slow scripts you should be aware that while the script is running no other client can execute commands.
Error handling
As already stated, calls to redis.call() resulting in a Redis command error will stop the execution of the script and return an error, in a way that makes it obvious that the error was generated by a script:
> del foo
(integer) 1
> lpush foo a
(integer) 1
> eval "return redis.call('get','foo')" 0
(error) ERR Error running script (call to f_6b1bf486c81ceb7edf3c093f4c48582e38c0e791): ERR Operation against a key holding the wrong kind of value
Using redis.pcall() no error is raised, but an error object is returned in the format specified above (as a Lua table with an err field). The script can pass the exact error to the user by returning the error object returned by redis.pcall().
Bandwidth and EVALSHA
The EVAL command forces you to send the script body again and again. Redis does not need to recompile the script every time as it uses an internal caching mechanism, however paying the cost of the additional bandwidth may not be optimal in many contexts.
```

**Comment by prdn** [comment]
@manast well my redis db that is around 1GB reached 15GB of memory, that's a pretty big overhead in my opinion and it is a risk on a production environment since there is no clear limit for it.

I think I found the issue:
https://github.com/OptimalBits/bull/blob/f1ea81c0e9a950f47d5aa39a2bc5ff82cb573355/lib/scripts.js#L328
Here you call redlock passing a variable locking Lua script.
This is probably what's bloating the redis cache

**Comment by manast** [comment]
without checking redlock, I guess that line is there because it is not currently possible to specify extra keys to be used by redlock. That would certainly need a change in redlock as well, to be merged and released before we can fix it in bull...

**Comment by prdn** [comment]
Sounds good. At least it seems the explanation of what I'm seeing. I'll see if I can prepare something for graylog. 
I think that any real use of bull-2.0 in production will end up in causing troubles. I'm handling around 1 million jobs per day and with the current cache grow I reach 15GB in 3 days.

Thanks for this awesome lib btw, I truly like it

**Comment by manast** [comment]
Even if the 15Gb is scary, it may not be a problem if redis just reuse that memory when needed. I checked redis.conf file but unfortunately there is no setting to limit the max amount of lua cache. Maybe https://github.com/antirez could give some explanation about this behaviour.

**Comment by prdn** [comment]
I agree that it would not be scary if there was a recycle of the memory. I checked for that parameter as well.
From here https://redis.io/commands/eval 
```The reason why scripts can be cached for long time is that it is unlikely for a well written application to have enough different scripts to cause memory problems. Every script is conceptually like the implementation of a new command, and even a large application will likely have just a few hundred of them. Even if the application is modified many times and scripts will change, the memory used is negligible.```

I have the feeling that the changing the script each time is against redis design.

**Comment by gblock0** [comment]
We are seeing this as well, our `used_memory_lua` was at `1118090240`. Our backend locks up and once we run `script flush` things start working again.

**Comment by joeandrews** [comment]
I think we have fixed this on our pull request.  https://github.com/OptimalBits/bull/blob/f1ea81c0e9a950f47d5aa39a2bc5ff82cb573355/lib/scripts.js#L328 was creating a new script for every single job ID. 10,000 jobs, results in 10,000 scripts in the script cache and hence this issue.

**Comment by jeltok** [comment]
We experience this issue as well with about 600k jobs a day. We use Redis 3.0.7 on Ubuntu 14.04

After 7 days of use Redis memory is arounf 8GB - data is less than 1GB, rest is reported as Lua scripts cache.

**Comment by prdn** [comment]
Any update on this issue? 
Thanks

**Comment by manast** [comment]
we have a PR that solves this issue but there is another issue with that PR that needs to be resolved before it can be merged.

**Comment by jeltok** [comment]
That's how the memory on my Redis instance (that supports Bull) looks like. We are running about 1 million messages a day through the Bull queues. The server has 16 GB of memory, and looses about 3GB a day to Lua cache.
Running SCRIPT FLUSH command any other day helps, but is a hustle. Other than that, Bull works just great for us.
Hope this issue will be fixed soon.

![image](https://cloud.githubusercontent.com/assets/1056801/23333060/d24c4b3c-fb52-11e6-8b7c-f8c577498c29.png)

**Comment by manast** [comment]
@jeltok a fix for this issue is the next thing in my list.

**Comment by manast** [comment]
@jeltok can you please verify that the issue is resolved with bull 2.2.0 so that we can close it?

**Comment by engvik** [comment]
Looks like we're still seeing the issue. Upgraded from 1.1.3 to 2.2.0 today. Have kept an eye on the `used_memory_lua` field and it's been steadily increasing since the upgrade. Flushing the scripts from the redis instance drops the usage down to normal.

![redis](https://cloud.githubusercontent.com/assets/851595/23479608/5710f712-fec5-11e6-82e6-7949dc3fc411.png)

**Comment by manast** [comment]
hmm, ok, I will need to dig more into it.

**Comment by engvik** [comment]
Thanks! Just switched from kue, and other than this it's been working great! Let me know if you need anything :smile:

**Comment by jeltok** [comment]
I have installed version 2.2.0 yesterday. Unfortunately it does not improve the issue with Lua cache by much.

![image](https://cloud.githubusercontent.com/assets/1056801/23511978/101c2abc-ff2d-11e6-92f0-e7700719c23f.png)

**Comment by manast** [comment]
@jeltok I would say nothing at all :).  I will try to do a deeper analysis.

**Comment by joeandrews** [comment]
@manast @jeltok I have some time today will try to look into it further as well.

Joe

**Comment by manast** [comment]
Ok, I have created a custom bull-redlock package and fixed one last issue. The queue works pretty well now without leaks. I will make a new release in the following hours.

**Comment by jeltok** [comment]
Thanks a lot!!! Looks much better now :-)

![image](https://cloud.githubusercontent.com/assets/1056801/23755464/68158a48-04ae-11e7-9c84-4b03059d5227.png)

**Comment by boatmeme** [comment]
Sorry for the noise, just wanted to send some gratitude to @manast for fixing this issue.
